compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -284,14 +328,35 @@ impl<'a> Parser<'a> {
             }
         } else if self.eat_keyword(kw::Impl) {
             self.parse_impl_ty(&mut impl_dyn_multi)?
+        } else if allow_anonymous_types == AllowAnonymousTypes::Yes
+            && self.can_start_anonymous_type()
+        {
+            self.parse_anonymous_ty(lo)?
         } else if self.is_explicit_dyn_type() {
             self.parse_dyn_ty(&mut impl_dyn_multi)?
         } else if self.eat_lt() {
             // Qualified path
             let (qself, path) = self.parse_qpath(PathStyle::Type)?;
             TyKind::Path(Some(qself), path)
         } else if self.check_path() {
-            self.parse_path_start_ty(lo, allow_plus, ty_generics)?
+            // `union` is a weak keyword, so it can possibly be a path
+            if allow_anonymous_types == AllowAnonymousTypes::OnlyRecover
+                && self.can_start_anonymous_type()
+            {
+                // If we can parse an anonymous type, do so and let
+                // ast validation error out later
+                let snapshot = self.create_snapshot_for_diagnostic();
+                match self.parse_anonymous_ty(lo) {
+                    Ok(ty) => ty,
+                    Err(err) => {
+                        err.cancel();
+                        self.restore_snapshot(snapshot);
+                        self.parse_path_start_ty(lo, allow_plus, ty_generics)?
+                    }
+                }
+            } else {
+                self.parse_path_start_ty(lo, allow_plus, ty_generics)?
+            }
```

Comment:
Ok, so this is the part of the change that I'm least sure of.  In this code, because `union` can be an identifier, you have to expressly opt-in to parsing an anonymous union type. `struct` on the other hand is a keyword and will attempt to parse an anonymous struct type for recovery. This is really more of hack than anything proper.

This has some implications, such as `impl struct {}` having an error mentioning anonymous struct types (instead of expected type found keyword), but `impl struct{} {}` having a "proper" error. This may require recovery changes in the impl block parsing beyond what I have already done, but I'm not too sure on this.

Reply:
@carbotaniuman:

@Amanieu and I were reviewing this, and wondering if it would be possible to confine this *entirely* to field definitions, and even then only when the field name is `_`.

If the field name is `_`, then look for `struct` or `union` in the type. If the type is `struct` then it *must* continue on with `{`. If the type is `union`, it *could* be `union;` or `union::something...`, either of which would be the *other* case of unnamed fields (where you have a named type), but if it isn't the start of a path then it has to continue with `{`.

Either way, we're wondering if this can *only* be parsed in field types, and nowhere else, avoiding the need to handle it (for recovery or otherwise) in other locations a type can appear.